templates/cluster-template-ipv6.yaml

```diff
@@ -0,0 +1,121 @@
+---
+apiVersion: cluster.x-k8s.io/v1beta1
```

Comment:
Q: Which test is using this template?
I don't see it added to e2e_conf.yaml for unmanaged tests.

Reply:
There are no unmanaged e2e tests for ipv6 because you can't make it IPv6 without manual mocking about.

Unless, you would want to implement installing a CNI, gathering the CIDR, passing that to the CNI config, updating the configmap, helm install calico, then terminate a pod and see if it comes back with ipv6. :D

I think that's a bit much. WDYT?

Reply:
You already tested it locally and things are working, do we have steps to make it work listed here in the doc?

> terminate a pod and see if it comes back with ipv6. :D

 Can't we start with IPv6? by passing a pre-defined CIDR to the CNI resource from the start?
Which configmap do we need to update?

Reply:
>Can't we start with IPv6? by passing a pre-defined CIDR to the CNI resource from the start?

By the time we get to this point the cluster is already there with a couple of nodes and pods running. So like, kube-proxy and such will already exist.

I have not found a way to install CNI without pods already existing. If CNI is not there, then the nodes can't join the cluster. They will keep being pending. I haven't found a way to install CNI first, with it controlling all pods even, kube-proxy pods. Maybe there is a way with CAPA to do that?

Also, another problem is that we don't know beforehand which subnet they node will be in. But it seems to work with the VPC defined CIDR anyways for now. With the restriction that calico, for example, only supports length of 112 and above for CIDR passed to pods. Which is probably still plenty, but still...

Reply:
BYOIP is possible in two ways. You either have one allocated by your ISP or your company, or, you create an AWS VPC with the setting `ipFamily` set to ipv6. In which case, AWS will give you an ipv6 CIDR.

If you have an existing CIDR you can provide that through the following settings:

```yaml
  clusterNetwork:
    pods:
      cidrBlocks: ["192.168.0.0/16", "2022:ff00::/56"]
    services:
      cidrBlocks: ["192.168.0.0/16", "2022:ff00::/56"]
...
spec:
  vpc:
    ipv6:
      ipv6CidrBlock: "2022:ff00::/56"
      ipv6PoolId: "pool-id"
```

Now that I see we don't need ipv6 prefix on the cidrblock anymore. :) heh.

If you have an existing network infrastructure with VPC and subnet configuration correctly laid out, you need to define the VPC ID and the subnets as you would normally when you bring your own infrastructure.

In that case, the CIDR is the VPC CIDR and you don't need to provide ipv6 details since CAPA will reconcile those.

I don't know how to do the rest for an unmanaged cluster. Like, CAPA doesn't install a CNI out of the box, right? If it does, how can I configure that?

Reply:
With ClusterResourceSet, once a single controlplane node comes up for the workload cluster, CNI is installed automatically.

Reply:
Even without existing network infra, with passing fd00:: a private CIDR, can this work>

Reply:
Oooh! Interesting! How does that work? Which CNI provider does it install?

Reply:
>Even without existing network infra, with passing fd00:: a private CIDR, can this work?

If the CNI can assign the right IP family and CIDR, then I suppose yes.

Reply:
Okay, I will test this and see with your branch.

ClusterResourceSet is not CNI specific, it just applies whatever objects user create as secret or configmap:

Reply:
https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20200220-cluster-resource-set.md

Reply:
Ahha, interesting. So, the calico configMap needs the following settings:

```
--cluster-pool-ipv6-cidr="2a05:d014:852:f02::/112"
--cluster-pool-ipv6-cidr-size="128"
```

But the problem here is that `2a05:d014:852:f02::/112` is specific to a subnet. So maybe try the VPC CIDR block but change the 56 to 112. And that should then work. Perhaps... I'm not sure and the calico documentation is attrocious. :(